lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs

```diff
@@ -413,6 +419,187 @@ spec = do
+                        case epochCeiling sps time of
+                            Nothing -> not withinBounds
+                            Just e1 -> case epochPred e1 of
+                                Nothing -> e1 == minBound
+                                Just e2 -> time > epochStartTime sps e2
+
+    describe "Epoch arithmetic: epochFloor: core properties" $ do
+
+        it "epochStartTime (epochFloor t) <= t" $
+            withMaxSuccess 1000 $ property $
+                \(SlotParametersAndTimePoint sps time) -> do
+                    let timeMinimum = epochStartTime sps minBound
+                    let withinBounds = time >= timeMinimum
+                    checkCoverage $
+                        cover 10      withinBounds  "within bounds" $
+                        cover 10 (not withinBounds) "out of bounds" $
+                        case epochFloor sps time of
+                            Nothing -> not withinBounds
+                            Just en -> time >= epochStartTime sps en
+
+        it "epochStartTime (epochSucc (epochFloor t)) > t" $
+            withMaxSuccess 1000 $ property $
+                \(SlotParametersAndTimePoint sps time) -> do
+                    let timeMinimum = epochStartTime sps minBound
+                    let withinBounds = time >= timeMinimum
+                    checkCoverage $
+                        cover 10      withinBounds  "within bounds" $
+                        cover 10 (not withinBounds) "out of bounds" $
+                        case epochFloor sps time of
+                            Nothing -> not withinBounds
+                            Just e1 -> case epochSucc e1 of
+                                Nothing -> e1 == maxBound
+                                Just e2 -> time < epochStartTime sps e2
+
+    describe "Epoch arithmetic: epochCeiling: boundary conditions" $ do
+
+        it "epochCeiling . epochStartTime == id" $
+            withMaxSuccess 1000 $ property $ \(sps, epoch) ->
+                epochCeiling sps (epochStartTime sps epoch)
+                    === Just epoch
+
+        it "epochCeiling . utcTimePred . epochStartTime == id" $
+            withMaxSuccess 1000 $ property $ \(sps, epoch) ->
+                epoch > minBound ==> do
+                    let fun = epochCeiling sps
+                            . utcTimePred
+                            . epochStartTime sps
+                    Just epoch === fun epoch
+
+        it "epochPred . epochCeiling . utcTimeSucc . epochStartTime == id" $
+            withMaxSuccess 1000 $ property $ \(sps, epoch) ->
+                epoch < maxBound ==> do
+                    let fun = (epochPred =<<)
+                            . epochCeiling sps
+                            . utcTimeSucc
+                            . epochStartTime sps
+                    Just epoch === fun epoch
+
+        it "epochCeiling (epochStartTime minBound) == minBound" $
```

Comment:
**[remark]**

This probably was already checked within `"epochCeiling . epochStartTime == id"` epoch generator as defined tends to pick low values presumably. Saying that, I am not against this stuff (if we do not introduce case coverage above)

Reply:
> This probably was already checked within "epochCeiling . epochStartTime == id" 

Well spotted!

I added these boundary cases in order to convince myself that they really were being tested.

Of course we could instead rely on the `epochCeiling . epochStartTime == id` test alone, and then check that all the cases are being covered.

I'll add this in a future PR.